### PR TITLE
Port image command to new cli structure

### PR DESCRIPTION
**Pull request by rugwirobaker**
* Port image command
* Add shared watch package

### CODE REVIEW

File: internal/cli/internal/watch/watch.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+	return g.Wait()
+}
+
+func renderLogs(ctx context.Context, alloc *api.AllocationStatus) {
+	out := iostreams.FromContext(ctx).Out
+	cfg := config.FromContext(ctx).JSONOutput
+
+	logPresenter := presenters.LogPresenter{
+		HideAllocID:    true,
```
Comment (rugwirobaker) [inline internal/cli/internal/watch/watch.go:270]: @azazeal do we have something in render to replace this logic?

File: internal/cli/internal/command/command.go [hunk 2]
```diff
@@ -426,3 +426,24 @@ func LoadAppNameIfPresent(ctx context.Context) (context.Context, error) {
 
 	return app.WithName(ctx, name), nil
 }
+
+func SwapWorkingDirectoryIfRequired(ctx context.Context) (context.Context, error) {
```
Comment (azazeal) [inline internal/cli/internal/command/command.go:430]: We must find a better name for this

File: internal/cli/internal/command/deploy/deploy.go [hunk 3]
```diff
@@ -151,7 +119,7 @@ func run(ctx context.Context) error {
 		tb := render.NewTextBlock(ctx, fmt.Sprintf("Release command detected: %s\n", releaseCommand.Command))
 		tb.Done("This release will not be available until the release command succeeds.")
 
-		if err := watchReleaseCommand(ctx, releaseCommand.ID); err != nil {
+		if err := watch.WatchReleaseCommand(ctx, releaseCommand.ID); err != nil {
```
Comment (azazeal) [inline internal/cli/internal/command/deploy/deploy.go:122]: `watch.WatchReleaseCommand` doesn't read all that great - maybe `watch.ReleaseCommand` would be better I think

File: internal/cli/internal/command/image/image.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"github.com/spf13/cobra"
+
+	"github.com/superfly/flyctl/internal/cli/internal/command"
+)
+
+func New() *cobra.Command {
+	const (
+		long  = "Manage app image"
```
Comment (azazeal) [inline internal/cli/internal/command/image/image.go:11]: `long` should always terminate with a line break.

you should:

```go
const (
    short  = "Manage app image"
    long = short + "\n"
)
```

File: internal/cli/internal/command/image/image.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+		short = "Manage app image"
+
+		usage = "image"
+	)
+
+	cmd := command.New(usage, short, long, nil)
+
+	cmd.Aliases = []string{"img"}
+
```
Comment (azazeal) [inline internal/cli/internal/command/image/image.go:20]: if this command takes no arguments, please set `cmd.Args` to `cobra.NoArgs`

File: internal/cli/internal/command/image/show.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/logrusorgru/aurora"
+	"github.com/spf13/cobra"
+
+	"github.com/superfly/flyctl/internal/cli/internal/app"
+	"github.com/superfly/flyctl/internal/cli/internal/command"
+	"github.com/superfly/flyctl/internal/cli/internal/config"
+	"github.com/superfly/flyctl/internal/cli/internal/flag"
+	"github.com/superfly/flyctl/internal/cli/internal/render"
+	"github.com/superfly/flyctl/internal/client"
+	"github.com/superfly/flyctl/pkg/iostreams"
```
Comment (azazeal) [inline internal/cli/internal/command/image/show.go:16]: `pkg` imports before the rest of our own, please

File: internal/cli/internal/command/image/show.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+func newShow() *cobra.Command {
+	const (
+		long  = "Show image details."
```
Comment (azazeal) [inline internal/cli/internal/command/image/show.go:21]: long should end with `\n`

File: internal/cli/internal/command/image/show.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+		if app.ImageDetails.Version != "" {
+			current = fmt.Sprintf("%s %s", current, app.ImageDetails.Version)
+		}
+
+		if app.LatestImageDetails.Version != "" {
+			latest = fmt.Sprintf("%s %s", latest, app.LatestImageDetails.Version)
+		}
+
+		fmt.Fprintln(io.ErrOut, aurora.Yellow(fmt.Sprintf("Update available! (%s -> %s)", current, latest)))
```
Comment (azazeal) [inline internal/cli/internal/command/image/show.go:70]: please print once if possible

File: internal/cli/internal/command/image/update.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	"github.com/superfly/flyctl/internal/cli/internal/render"
+	"github.com/superfly/flyctl/internal/cli/internal/watch"
+	"github.com/superfly/flyctl/internal/client"
+)
+
+func newUpdate() *cobra.Command {
+	const (
+		long = `This will update the application's image to the latest available version.
+The update will perform a rolling restart against each VM, which may result in a brief service disruption.
+		`
```
Comment (azazeal) [inline internal/cli/internal/command/image/update.go:26]: the closing backtick shouldn't be indented.

File: internal/cli/internal/command/image/update.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+		short = "Updates the app's image to the latest available version. (Fly Postgres only)"
+
+		usage = "update"
+	)
+
+	cmd := command.New(usage, short, long, runUpdate,
+		command.RequireSession,
+		command.RequireAppName,
+	)
+
```
Comment (azazeal) [inline internal/cli/internal/command/image/update.go:36]: if this command takes no args, use `cobra.NoArgs` please

File: internal/cli/internal/command/image/update.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+func runUpdate(ctx context.Context) error {
+	var (
+		client  = client.FromContext(ctx).API()
+		io      = iostreams.FromContext(ctx)
+		appName = app.NameFromContext(ctx)
+	)
+
+	app, err := client.GetImageInfo(ctx, appName)
+	if err != nil {
+		return err
```
Comment (azazeal) [inline internal/cli/internal/command/image/update.go:59]: there should be a decoration here: `fmt.Errorf("failed retrieving image info: %w", err)`

File: internal/cli/internal/command/image/update.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	release, releaseCommand, err := client.DeployImage(ctx, input)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(io.Out, "Release v%d created\n", release.Version)
+	if releaseCommand != nil {
+		fmt.Fprintf(io.Out, "Release command detected: this new release will not be available until the command succeeds.\n")
```
Comment (azazeal) [inline internal/cli/internal/command/image/update.go:105]: `You want `Fprintln` here

File: internal/cli/internal/watch/watch.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+	"github.com/superfly/flyctl/internal/cli/internal/app"
+	"github.com/superfly/flyctl/internal/cli/internal/config"
+	"github.com/superfly/flyctl/internal/cli/internal/format"
+	"github.com/superfly/flyctl/internal/cli/internal/render"
+	"github.com/superfly/flyctl/internal/client"
+	"github.com/superfly/flyctl/internal/deployment"
+	"github.com/superfly/flyctl/internal/flyerr"
+)
+
+func WatchDeployment(ctx context.Context) error {
```
Comment (azazeal) [inline internal/cli/internal/watch/watch.go:30]: should just be named `Deployment`

File: internal/cli/internal/watch/watch.go [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+	if !monitor.Success() {
+		tb.Done("Troubleshooting guide at https://fly.io/docs/getting-started/troubleshooting/")
+		return flyerr.ErrAbort
+	}
+
+	return nil
+}
+
+func WatchReleaseCommand(ctx context.Context, id string) error {
```
Comment (azazeal) [inline internal/cli/internal/watch/watch.go:149]: should just be named `ReleaseCommand`

File: internal/cli/internal/watch/watch.go (hunk 1 above)
Comment (azazeal) [reply internal/cli/internal/watch/watch.go:270]: Nope, you should implement something as we don't use `presenters' any more

**Review by azazeal — CHANGES_REQUESTED** [review]
first pass

File: internal/cli/internal/command/command.go (hunk 2 above)
Comment (rugwirobaker) [reply internal/cli/internal/command/command.go:430]: ChangeWorkDirIfRequired?

**Comment by azazeal** [comment]
I think the name should clarify that it works with the first arg or
something like that.

File: internal/cli/internal/command/command.go [hunk 15]
```diff
@@ -426,3 +426,24 @@ func LoadAppNameIfPresent(ctx context.Context) (context.Context, error) {
 
 	return app.WithName(ctx, name), nil
 }
+
+func WorkingDirIsFirsArg(ctx context.Context) (context.Context, error) {
```
Comment (azazeal) [inline internal/cli/internal/command/command.go:430]: `ChangeWorkingDirectoryToFirstArgIfPresent` should be more descriptive.

File: internal/cli/internal/command/image/update.go (hunk 10 above)
Comment (azazeal) [reply internal/cli/internal/command/image/update.go:36]: Does this command take args or not?

File: internal/cli/internal/render/logs.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+func LogEntry(w io.Writer, entry logs.LogEntry, opts ...LogOption) (err error) {
+	options := &LogOptions{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if options.JSONFormat {
+		outBuf, _ := json.MarshalIndent(entry, "", "    ")
```
Comment (azazeal) [inline internal/cli/internal/render/logs.go:61]: don't we have `JSON` defined in this package already?

**Review by azazeal — APPROVED** [review]
There's a lot of meat in this PR. Some things would probably need some further touch-up in the near future.

**Pull request merged**